drivers/pppos.c

```diff
@@ -440,6 +463,12 @@ static void pppos_mainLoop(void* _state)
 	int running = 1;
 
 	while (running) {
+		mutexLock(state->lock);
+		while (!(state->ppp->netif->flags & NETIF_FLAG_UP)) {
```

Comment:
**This works just by accident.**

The `netif->flags` is not a volatile field, so compiler may optimize it out (keep the value in register).

We should probably add `volatile int want_connected;` to `state` struct and make decision based on that.

side note: `netif` is available directly in `state`.

Reply:
Right, I have added the volatile member. The problem should be solved now.